src/storage/backend/sqlite/SQLite.cc

```diff
@@ -279,36 +313,113 @@ OperationResult SQLite::DoErase(ResultCallback* cb, ValPtr key) {
     if ( ! key_data )
         return {ReturnCode::SERIALIZATION_FAILED, "Failed to serialize key"};
 
-    auto stmt = erase_stmt.get();
+    auto stmt = unique_stmt_ptr(erase_stmt.get(), sqlite3_reset);
 
-    if ( auto res = CheckError(sqlite3_bind_blob(stmt, 1, key_data->data(), key_data->size(), SQLITE_STATIC));
+    if ( auto res = CheckError(sqlite3_bind_blob(stmt.get(), 1, key_data->data(), key_data->size(), SQLITE_STATIC));
          res.code != ReturnCode::SUCCESS ) {
-        sqlite3_reset(stmt);
         return res;
     }
 
-    return Step(stmt, false);
+    return Step(stmt.get(), false);
 }
 
 /**
  * Removes any entries in the backend that have expired. Can be overridden by
  * derived classes.
  */
 void SQLite::DoExpire(double current_network_time) {
-    auto stmt = expire_stmt.get();
+    int status;
+    char* errMsg = nullptr;
+    unique_stmt_ptr stmt;
+
+    // Begin an exclusive transaction here to lock the database for this one process. That
+    // will ensure there isn't a TOCTOU bug in the time check below.
+    status = SQLITE_BUSY;
+    while ( status == SQLITE_BUSY ) {
+        status = sqlite3_exec(db, "begin exclusive transaction", nullptr, nullptr, &errMsg);
+        if ( status == SQLITE_OK )
+            break;
+        else if ( status == SQLITE_BUSY )
+            // I honestly don't know if we need to wait here. This should fall under the
+            // 'busy_timeout' pragma timeout, which means that we've already waited for
+            // some amount of time before even getting here.
+            std::this_thread::sleep_for(5ms);
```

Comment:
I don't know if this even needs to wait and try again at all. If another process is holding the transaction, it means they're handling expiration already. It might be fine for the other processes to just return here.